src/pudl/analysis/record_linkage/embed_dataframe.py

```diff
@@ -0,0 +1,181 @@
+class FeatureMatrix:
+    """Class to wrap a feature matrix returned from dataframe embedding.
+
+    Depending on the transformations applied, a feature matrix may be sparse or dense
+    matrix. Using this wrapper enables Dagsters type checking while allowing both dense
+    and sparse matrices underneath.
+    """
+
+    matrix: np.ndarray | scipy.sparse.csr_matrix
+
+
+class TransformStep(BaseModel, ABC):
+    """TransformStep's can be combined to vectorize one or more columns.
+
+    This class defines a very simple interface for TransformStep's, which essentially
+    says that a TransformStep should take configuration and implement the method as_transformer.
+    """
+
+    name: str
+
+    @abstractmethod
+    def as_transformer(self) -> BaseEstimator:
+        """This method should use configuration to produce a :class:`sklearn.base.BaseEstimator`."""
+        ...
+
+
+class ColumnVectorizer(BaseModel):
+    """Define a set of transformations to apply to one or more columns."""
+
+    transform_steps: list[TransformStep]
+    weight: float = 1.0
+    columns: list[str]
+
+    def as_pipeline(self):
+        """Return :class:`sklearn.pipeline.Pipeline` with configuration."""
+        return Pipeline(
+            [
+                (
+                    step.name,
+                    step.as_transformer(),
+                )
+                for step in self.transform_steps
+            ]
+        )
+
+
+def dataframe_embedder_factory(vectorizers: dict[str, ColumnVectorizer]):
+    """Return a configured op graph to embed an input dataframe."""
+
+    @op
+    def train_dataframe_embedder(df: pd.DataFrame):
+        """Train :class:`sklearn.compose.ColumnTransformer` on input."""
+        column_transformer = ColumnTransformer(
+            transformers=[
+                (name, column_transform.as_pipeline(), column_transform.columns)
+                for name, column_transform in vectorizers.items()
+            ],
+            transformer_weights={
+                name: column_transform.weight
+                for name, column_transform in vectorizers.items()
+            },
+        )
+
+        return column_transformer.fit(df)
+
+    @op
+    def apply_dataframe_embedder(df: pd.DataFrame, transformer: ColumnTransformer):
+        """Use :class:`sklearn.compose.ColumnTransformer` to transform input."""
+        return FeatureMatrix(matrix=transformer.transform(df))
+
+    @graph
+    def embed_dataframe(df: pd.DataFrame) -> FeatureMatrix:
+        """Train dataframe embedder and apply to input df."""
+        transformer = train_dataframe_embedder(df)
+        return apply_dataframe_embedder(df, transformer)
+
+    return embed_dataframe
+
+
+class TextVectorizer(TransformStep):
+    """Implement TransformStep for :class:`sklearn.feature_extraction.text.TfidfVectorizer`."""
+
+    name: str = "tfidf_vectorizer"
+
+    #: See sklearn documentation for all options
+    options: dict = {"analyzer": "char", "ngram_range": (2, 10)}
```

Comment:
I was kind of viewing these as our defaults, and intentionally not mirroring `sklearn`. Particularly if these values work for the ferc-eia match, I think it would be good to use our own set of defaults.